### PR TITLE
fix(sdk,shared): separate approvals from smart contract calls [WEB-299]

### PR DESCRIPTION
**Pull request by acdibble**

### CODE REVIEW

File: packages/sdk/src/swap/sdk.ts [hunk 1]
```diff
@@ -72,9 +74,22 @@ export class SwapSDK {
 
   executeSwap(params: ExecuteSwapParams): Promise<ContractReceipt> {
     assert(this.signer, 'No signer provided');
-    return executeSwap(params, {
-      network: this.network,
+    return executeSwap(params, { network: this.network, signer: this.signer });
```
Comment (niklasnatter) [inline packages/sdk/src/swap/sdk.ts:77]: would prefer returning the hash here too to make things consistent with the other methods

File: packages/sdk/src/swap/sdk.ts [hunk 2]
```diff
@@ -72,9 +74,22 @@ export class SwapSDK {
...
+  }
+
+  async approveVault(
```
Comment (niklasnatter) [inline packages/sdk/src/swap/sdk.ts:80]: should we expose a method like `needsVaultApproval`?

File: packages/sdk/src/funding/index.ts [hunk 3]
```diff
@@ -64,31 +60,22 @@ export class FundingSDK {
...
-    return balance.toBigInt();
+    return getFlipBalance(this.options.network, this.options.signer);
   }
 
   /**
    * @param amount the amount of FLIP to request approval for
    * @returns the transaction hash or null if no approval was required
    */
-  async requestFlipApproval(
+  async approveStateChainGateway(
```
Comment (niklasnatter) [inline packages/sdk/src/funding/index.ts:70]: how do we determine if we need an approval in the auctions app? should we expose a method like `needsStateChainGatewayApproval`?

File: packages/shared/src/consts.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import { Asset, Chain, Chains, Assets, ChainflipNetworks } from './enums';
+
+// !!!!! IMPORTANT !!!!!
+// Do not change these indices.
+export const chainMap: Record<Chain, number> = {
```
Comment (niklasnatter) [inline packages/shared/src/consts.ts:5]: `chainContractIds`?

File: packages/shared/src/contracts.ts [hunk 5]
```diff
@@ -53,22 +29,31 @@ export const getStateChainGatewayContractAddress = (
...
+  erc20: ERC20,
+  allowance: BigNumberish,
+  nonce?: bigint | number | string,
+): Promise<ContractReceipt | null> => {
+  const amountBigNumber = BigNumber.from(amount);
+  const allowanceBigNumber = BigNumber.from(allowance);
+  if (allowanceBigNumber.gte(amountBigNumber)) return null;
+  const requiredAmount = amountBigNumber.sub(allowanceBigNumber);
+  const tx = await erc20.approve(spenderAddress, requiredAmount, { nonce });
+  return tx.wait(1);
```
Comment (niklasnatter) [inline packages/shared/src/contracts.ts:56]: think we should try to handle transactions consistent - either we should always wait for inclusion before returning or return immediately after submitting

Comment (acdibble) [reply packages/shared/src/contracts.ts:56]: I don't know anywhere we don't wait for inclusion before returning a hash to the consumer

File: packages/shared/src/vault/approval.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  return checkAllowance(
+    params.amount,
+    vaultContractAddress,
+    erc20Address,
+    opts.signer,
+  );
+};
+
+export const approveVault = async (
+  params: TokenSwapParams,
```
Comment (albert-llimos) [inline packages/shared/src/vault/approval.ts:37]: Does it make sense to have this TokenSwapParams here? That contains destination parameters which won't be used here. Might be weird/misleading in the SDK.

Comment (acdibble) [reply packages/shared/src/vault/approval.ts]: Fixed

File: packages/sdk/src/swap/sdk.ts (hunk 2 above)
Comment (zoheb391) [reply packages/sdk/src/swap/sdk.ts:85]: +1 but could be its own ticket.

**Review by zoheb391 — APPROVED** [review]

**Review by albert-llimos — APPROVED** [review]
I tested it with the bouncer and it fixed the issue :rocket:

**Pull request merged**